lib/puppet/file_system.rb

```diff
@@ -348,6 +349,21 @@ def self.path_string(path)
     @impl.path_string(path)
   end
 
+  # Produces a string representation of the opaque path handle, with expansions
+  # performed on ~.  For Windows, this means that C:\Users\Admini~1\AppData will
+  # be expanded to C:\Users\Administrator\AppData.  On POSIX filesystems, the
+  # value ~ will be expanded to something like /Users/Foo
+  #
+  # This method exists primarlily to resolve a Ruby deficiency where
+  # File.expand_path doesn't handle ~ in each segment on a Windows path
+  #
+  # @param path [Object] a path handle produced by {#pathname}
+  # @return [String] a string representation of the path
+  #
+  def self.expanded_path_string(path)
```

Comment:
I've thought about this a bit more and I disagree, because I think we ultimately want a corrected `expand_path` for the sake of Windows (just as we have fixed `stat` and `lstat` methods for Windows in our `FileSystem` abstraction).  I think the best API here is to follow the semantics of `expand_path` rather than some weird / confusing in-between thing with only one purpose.

Then we can just choose at the call site if we want to expand or not - which pushes complexity out a little bit for the sake of keeping a more reasonably `FileSystem` API.  So it may be that we just don't call it on 'nix for Environment / Module loading... but my latest example might not carry as much of a penalty given the path is now adjusted at the highest Environment level.

I'll have to run a few more perf tests to prove that out, but things are looking good on Windows again.